ports/raspberrypi/common-hal/rp2pio/StateMachine.c

```diff
@@ -1036,123 +1060,331 @@ uint8_t rp2pio_statemachine_program_offset(rp2pio_statemachine_obj_t *self) {
+    rp2pio_statemachine_clear_dma_write(pio_index, sm);
+    memset(&self->current_write_buf, 0, sizeof(self->current_write_buf));
+    memset(&self->next_write_buf_1, 0, sizeof(self->next_write_buf_1));
+    memset(&self->next_write_buf_2, 0, sizeof(self->next_write_buf_2));
+    memset(&self->next_write_buf_3, 0, sizeof(self->next_write_buf_3));
+    self->pending_buffers_write = 0;
     return true;
 }
 
 bool common_hal_rp2pio_statemachine_get_writing(rp2pio_statemachine_obj_t *self) {
-    return !self->dma_completed;
+    return !self->dma_completed_write;
+}
+
+int common_hal_rp2pio_statemachine_get_pending_write(rp2pio_statemachine_obj_t *self) {
+    return self->pending_buffers_write;
+}
+
+// =================================================================================
+
+bool common_hal_rp2pio_statemachine_background_read(rp2pio_statemachine_obj_t *self, uint8_t stride_in_bytes, bool swap) {
+
+    uint8_t pio_index = pio_get_index(self->pio);
+    uint8_t sm = self->state_machine;
+
+    self->switched_read_buffers = false;
+
+    int pending_buffers_read = (self->once_read_buf_info.info.len != 0) + (self->loop_read_buf_info.info.len != 0) + (self->loop2_read_buf_info.info.len != 0);
+
+    // If all buffer arguments have nonzero length, read once_read_buf, loop_read_buf, loop2_read_buf and repeat last two forever
+
+    if (!(self->once_read_buf_info.info.len)) {
+        if (!(self->loop_read_buf_info.info.len)) {
+            // If once_read_buf and loop_read_buf have zero length, read loop2_read_buf forever
+            self->once_read_buf_info = self->loop2_read_buf_info;
+            self->loop_read_buf_info = self->loop2_read_buf_info;
+        } else {
+            if (!(self->loop2_read_buf_info.info.len)) {
+                // If once_read_buf and loop2_read_buf have zero length, read loop_read_buf forever
+                self->once_read_buf_info = self->loop_read_buf_info;
+                self->loop2_read_buf_info = self->loop_read_buf_info;
+            } else {
+                // If only once_read_buf has zero length, read loop_read_buf, loop2_read_buf, and repeat last two forever
+                self->once_read_buf_info = self->loop_read_buf_info;
+                self->loop_read_buf_info = self->loop2_read_buf_info;
+                self->loop2_read_buf_info = self->once_read_buf_info;
+            }
+        }
+    } else {
+        if (!(self->loop_read_buf_info.info.len)) {
+            // If once_read_buf has nonzero length and loop_read_buf has zero length, read once_read_buf, loop2_read_buf and repeat last buf forever
+            self->loop_read_buf_info = self->loop2_read_buf_info;
+        } else {
+            if (!(self->loop2_read_buf_info.info.len)) {
+                // If once_read_buf has nonzero length and loop2_read_buf have zero length, read once_read_buf, loop_read_buf and repeat last buf forever
+                self->loop2_read_buf_info = self->loop_read_buf_info;
+
+            }
```

Comment:
it feels like the handling of "once buf and two loop bufs" could be factored out into shared code. However, I'm not asking you to do this right now.